src/shared/bg_public.h

```diff
@@ -1525,6 +1525,159 @@ meansOfDeath_t            BG_MeansOfDeathByName( const char *name );
+		}
+		count_ = other.count_;
+		return *this;
+	}
+	~BoundedVector() {
+		while (pop_back());
+	}
+
+	class iterator
+	{
+	private:
+		BoundedVector<T, maximum_size_>* vector_;
+		size_t index_;
+	public:
+		// Previously provided by std::iterator - see update below
+		using value_type        = T;
+		using difference_type   = std::ptrdiff_t;
+		using pointer           = T*;
+		using reference         = T&;
+		using iterator_category = std::random_access_iterator_tag;
+
+		iterator()
+			: vector_(nullptr), index_(0) {}
+		iterator(const iterator &other)
+			: vector_(other.vector_), index_(other.index_) {}
+		iterator(BoundedVector<T, maximum_size_> *vector, size_t index)
+			: vector_(vector), index_(index) {}
+		iterator& operator=(const iterator &other) {
+			vector_ = other.vector_;
+			index_ = other.index_;
+			return *this;
+		}
+		T& operator*() const { return vector_->array[index_]; }
+		bool operator==(const iterator& other) const {
+			return index_ == other.index_;
+		}
+		bool operator!=(const iterator& other) const {
+			return !(*this == other);
+		}
+		bool operator<(const iterator &other) const {
+			return index_ < other.index_;
+		}
+		bool operator<=(const iterator &other) const {
+			return index_ <= other.index_;
+		}
+		bool operator>(const iterator &other) const {
+			return index_ > other.index_;
+		}
+		bool operator>=(const iterator &other) const {
+			return index_ >= other.index_;
+		}
+		iterator operator++(int) {
+			iterator ret(vector_, index_);
+			++*this;
+			return ret;
+		}
+		iterator& operator++() {
+			++index_;
+			return *this;
+		}
+		iterator operator--(int) {
+			iterator ret(vector_, index_);
+			--*this;
+			return ret;
+		}
+		iterator& operator--() {
+			--index_;
+			return *this;
+		}
+		iterator operator-(int diff) const {
+			return iterator(vector_, index_ - diff);
+		}
+		iterator operator+(int diff) const {
+			return iterator(vector_, index_ + diff);
+		}
+		int operator-(const iterator &other) const {
+			ASSERT_EQ(vector_, other.vector_);
+			return (int)other.index_ - (int)index_;
+		}
+		int operator+(const iterator &other) const {
+			ASSERT_EQ(vector_, other.vector_);
+			return other.index_ + index_;
+		}
+		iterator& operator+=(size_t diff) {
+			index_ += diff;
+			return *this;
+		}
+		iterator& operator-=(size_t diff) {
+			index_ -= diff;
+			return *this;
+		}
+		T& operator[](int diff) const {
+			return vector_->array[index_ + diff];
+		}
+		void swap(iterator &other) {
+			std::swap(other.vector_, vector_);
+			std::swap(other.index_, index_);
+		}
+	};
+
+	T& operator[](size_t n) {
+		ASSERT_LT(n, count_);
+		return array[n];
+	}
+	Util::optional<T> pop_back() {
+		if (count_ == 0)
+			return {};
+		return std::move(array[--count_]);
+	}
+	void clear() {
+		while (pop_back()) {}
+	}
+	bool append(T elem) {
+		if (count_ == maximum_size_)
+			return false;
+		array[count_++] = std::move(elem);
+		return true;
+	}
+	size_t count() const { return count_; }
```

Comment:
This should be called `size`

Reply:
I'm not fond of calling it size, as the size of the vector is constant (in that it always takes as much memory space).

Reply:
updated